### PR TITLE
remove Mypy errors with new version of Mypy

### LINKED ISSUE #2046: Mypy Errors with new version of mypy
## 🐛 Bug description
With the new version of mypy, ignite/handlers/checkpoint.py and ignite/utils.py has some errors related to typing.
```
ignite/utils.py:161: error: unused "type: ignore" comment
        existing = name is None or name in logging.root.manager.loggerDict...
        ^
ignite/handlers/checkpoint.py:9: error: unused "type: ignore" comment
    from tempfile import _TemporaryFileWrapper  # type: ignore[attr-define...
    ^
ignite/handlers/checkpoint.py:695: error: Incompatible types in assignment
(expression has type "None", variable has type "_TemporaryFileWrapper") 
[assignment]
                tmp = None  # type: _TemporaryFileWrapper
                      ^
ignite/handlers/checkpoint.py:697: error: Incompatible types in assignment
(expression has type "IO[bytes]", variable has type "_TemporaryFileWrapper") 
[assignment]
                    tmp = tempfile.NamedTemporaryFile(delete=False, dir=se...
                          ^
Found 4 errors in 2 files (checked 98 source files)
```

## Environment

- Ignite Version (e.g., 0.3.0): Current master branch
- OS (e.g., Linux): Linux
- How you installed Ignite (`conda`, `pip`, source): source
- Python version: 3.8.6
- Any other relevant information:
    mypy version: mypy-0.901-cp38-cp38-manylinux2010_x86_64

### PR DESCRIPTION
**Pull request by sahilg06**
Fixes #2046 

Description: With the new version of mypy, ignite/handlers/checkpoint.py and ignite/utils.py had some errors related to typing.

Check list:

- [x] removed the errors

### CODE REVIEW

File: ignite/handlers/checkpoint.py [hunk 1]
```diff
@@ -692,10 +692,10 @@ def _save_func(self, checkpoint: Mapping, path: str, func: Callable, rank: int =
         else:
             tmp_file = None
             tmp_name = ""
-            tmp = None  # type: _TemporaryFileWrapper
+            tmp = None 
             if rank == 0:
                 tmp = tempfile.NamedTemporaryFile(delete=False, dir=self.dirname)
-                tmp_file = tmp.file
+                tmp_file = tmp.file # type: ignore
```
[the remark this answers is not in the document]
Comment (sahilg06) [reply ignite/handlers/checkpoint.py:698]: @vfdev-5 before ignoring this type check I was getting this error
```
 error: "IO[bytes]" has no attribute "file"; maybe "fileno"?
```

File: ignite/handlers/checkpoint.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 import warnings
 from abc import ABCMeta, abstractmethod
 from collections import OrderedDict
-from tempfile import _TemporaryFileWrapper  # type: ignore[attr-defined]
+from tempfile import _TemporaryFileWrapper
```
[the remark this answers is not in the document]
Comment (trsvchn) [reply ignite/handlers/checkpoint.py:9]: Yes if we remove it from line 695

File: ignite/handlers/checkpoint.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (trsvchn) [reply ignite/handlers/checkpoint.py:698]: I think I added this as a workaround
Maybe we need to refactor it? @vfdev-5

File: ignite/utils.py [hunk 3]
```diff
@@ -158,7 +158,7 @@ def setup_logger(
         Added ``reset`` parameter.
     """
     # check if the logger already exists
-    existing = name is None or name in logging.root.manager.loggerDict  # type: ignore
+    existing = name is None or name in logging.root.manager.loggerDict
```
Comment (trsvchn) [inline ignite/utils.py:161]: This seems OK

File: ignite/handlers/checkpoint.py [hunk 4]
```diff
@@ -692,10 +692,10 @@ def _save_func(self, checkpoint: Mapping, path: str, func: Callable, rank: int =
         else:
             tmp_file = None
             tmp_name = ""
-            tmp = None  # type: _TemporaryFileWrapper
+            tmp = None
```
Comment (trsvchn) [inline ignite/handlers/checkpoint.py:695]: This
```suggestion
            tmp : Optional[IO[bytes]] = None
```
with 

```python
                tmp_file = tmp.file # type: ignore
```

Should work

**Review by trsvchn — CHANGES_REQUESTED** [review]
Thanks for PR @sahilg06 !
Try to implement suggested changes! I hope this helps

File: ignite/handlers/checkpoint.py (hunk 4 above)
Comment (sahilg06) [reply ignite/handlers/checkpoint.py:695]: @trsvchn but IO is not defined anywhere

Comment (trsvchn) [reply ignite/handlers/checkpoint.py:695]: Yes, import it from typing, like we did for Optional etc

Comment (sahilg06) [reply ignite/handlers/checkpoint.py:695]: Yes got it. Thanks a lot @trsvchn

**Comment by trsvchn** [comment]
@sahilg06 Thanks for the updates

But checks failed due to code style issues. See here:

https://github.com/pytorch/ignite/runs/2805100275#step:8:42

could you please run code formatters (in your case isort)

you can find detailed steps in our contributing guide,

https://github.com/pytorch/ignite/blob/master/CONTRIBUTING.md#formatting-code

In addition, you can turn on code-style workflow in your local repo fork in "Actions" tab

**Comment by trsvchn** [comment]
@sahilg06 Thanks for the fix, I've noticed that in `requirements.txt` you use old verion of mypy, the main aim is to use the latest version, so let's try this pr vs latest mypy version

```
#mypy
mypy==0.812
```

**Comment by sahilg06** [comment]
> @sahilg06 Thanks for the fix, I've noticed that in `requirements.txt` you use old verion of mypy, the main aim is to use the latest version, so let's try this pr vs latest mypy version
> 
> ```
> #mypy
> mypy==0.812
> ```

Yes I upgraded the version while running the mypy tests on my local repo.

**Comment by trsvchn** [comment]
> > @sahilg06 Thanks for the fix, I've noticed that in `requirements.txt` you use old verion of mypy, the main aim is to use the latest version, so let's try this pr vs latest mypy version
> > ```
> > #mypy
> > mypy==0.812
> > ```
> 
> Yes, I upgraded the version while running the mypy tests on my local repo.

That's great! Now we have to do the same here to make CI happy :)

**Comment by sahilg06** [comment]
Are these above tests failing because of an older version of mypy? @trsvchn  @vfdev-5

**Comment by vfdev-5** [comment]
@sahilg06 yes, please remove what was added here : https://github.com/pytorch/ignite/commit/86cec1165441a945c93c938c23544ac52a17a170

mypy version should be latest stable

**Comment by sahilg06** [comment]
> @sahilg06 yes, please remove what was added here : [86cec11](https://github.com/pytorch/ignite/commit/86cec1165441a945c93c938c23544ac52a17a170)
> 
> mypy version should be latest stable

Shall I also add the newer version or Will removing those 2 lines be fine?

**Comment by vfdev-5** [comment]
Just remove the last line and uncomment commented line.

@ydcjeff @trsvchn  seems this line is not used ?
https://github.com/pytorch/ignite/blob/master/tests/run_code_style.sh#L15

**Comment by trsvchn** [comment]
> Just remove the last line and uncomment commented line.
> 
> @ydcjeff @trsvchn seems this line is not used ?
> https://github.com/pytorch/ignite/blob/master/tests/run_code_style.sh#L15

in `unit-tests` it is used here:

https://github.com/pytorch/ignite/blob/86cec1165441a945c93c938c23544ac52a17a170/.github/workflows/unit-tests.yml#L85

**Comment by trsvchn** [comment]
@vfdev-5 Oh, I got it!

**Comment by trsvchn** [comment]
I think you need to
```bash
pip install mypy -U
```

after the first install of mypy==812

**Comment by sahilg06** [comment]
> I think you need to
> 
> ```shell
> pip install mypy -U
> ```
> 
> after the first installation of mypy==812

So should requirements-dev.txt  be like this?
# mypy
mypy==0.812
mypy -U

**Comment by trsvchn** [comment]
> > I think you need to
> > ```shell
> > pip install mypy -U
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > after the first installation of mypy==812
> 
> So should requirements-dev.txt be like this?
> 
> ```
> # mypy
> mypy==0.812
> mypy -U
> ```
> 
> or simply
> 
> ```
> mypy
> ```

No, no, `requirements-dev.txt ` is ok. its all about how to get the latest mypy veriosn after executing this command

```bash
pip install -r requirements-dev.txt
```

**Comment by vfdev-5** [comment]
Here we fix the version with requirements-dev.txt and later we have to make things uniform between this file and run_code_style.sh

**Comment by ydcjeff** [comment]
We don't have [`mypy` in `requirements-dev.txt`](https://github.com/pytorch/ignite/blob/f5b0c5cb17cdc2a57a7a6790038e50ee652b94b2/requirements-dev.txt) until 86cec1165441a945c93c938c23544ac52a17a170. So `mypy` can be removed from `requirements-dev.txt`.

**Comment by trsvchn** [comment]
Now I see, yes I think we can remove `mypy` from `requirements-dev.txt`
@vfdev-5 ?

**Comment by sahilg06** [comment]
> We don't have [`mypy` in `requirements-dev.txt`](https://github.com/pytorch/ignite/blob/f5b0c5cb17cdc2a57a7a6790038e50ee652b94b2/requirements-dev.txt) until [86cec11](https://github.com/pytorch/ignite/commit/86cec1165441a945c93c938c23544ac52a17a170). So `mypy` can be removed from `requirements-dev.txt`.

Yes, its already present in tests/run_code_style.sh.

**Review by trsvchn — APPROVED** [review]
Thanks @sahilg06 ! LGTM!

**Review by vfdev-5 — APPROVED** [review]
Thanks @sahilg06 !

**Pull request merged**